indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/CostBasedAutoScaler.java

```diff
@@ -0,0 +1,310 @@
+  private final ScheduledExecutorService scalingDecisionExec;
+  private final SupervisorSpec spec;
+  private final SeekableStreamSupervisor supervisor;
+  private final CostBasedAutoScalerConfig config;
+  private final ServiceEmitter emitter;
+  private final ServiceMetricEvent.Builder metricBuilder;
+  private final WeightedCostFunction costFunction;
+
+  public CostBasedAutoScaler(
+      SeekableStreamSupervisor supervisor,
+      String dataSource,
+      CostBasedAutoScalerConfig config,
+      SupervisorSpec spec,
+      ServiceEmitter emitter
+  )
+  {
+    this.dataSource = dataSource;
+    this.config = config;
+    this.spec = spec;
+    this.supervisor = supervisor;
+    this.emitter = emitter;
+
+    final String supervisorId = StringUtils.format("Supervisor-%s", dataSource);
+
+    this.currentMetrics = new AtomicReference<>(null);
+    this.costFunction = new WeightedCostFunction();
+
+    this.metricsCollectionExec = Execs.scheduledSingleThreaded(
+        StringUtils.encodeForFormat(supervisorId) + "-CostBasedMetrics-%d"
+    );
+    this.scalingDecisionExec = Execs.scheduledSingleThreaded(
+        StringUtils.encodeForFormat(supervisorId) + "-CostBasedScaling-%d"
+    );
+
+    this.metricBuilder = ServiceMetricEvent.builder()
+                                           .setDimension(DruidMetrics.DATASOURCE, dataSource)
+                                           .setDimension(
+                                               DruidMetrics.STREAM,
+                                               this.supervisor.getIoConfig().getStream()
+                                           );
+  }
+
+  @Override
+  public void start()
+  {
+    Callable<Integer> scaleAction = () -> computeOptimalTaskCount(currentMetrics);
+    Runnable onSuccessfulScale = () -> currentMetrics.set(null);
+
+    metricsCollectionExec.scheduleAtFixedRate(
+        collectMetrics(),
+        config.getMetricsCollectionIntervalMillis(),
+        config.getMetricsCollectionIntervalMillis(),
+        TimeUnit.MILLISECONDS
+    );
+
+    scalingDecisionExec.scheduleAtFixedRate(
+        supervisor.buildDynamicAllocationTask(scaleAction, onSuccessfulScale, emitter),
+        config.getScaleActionStartDelayMillis(),
+        config.getScaleActionPeriodMillis(),
+        TimeUnit.MILLISECONDS
+    );
+
+    log.info(
+        "CostBasedAutoScaler started for dataSource [%s]: collecting metrics every [%d]ms, "
+        + "evaluating scaling every [%d]ms",
+        dataSource,
+        config.getMetricsCollectionIntervalMillis(),
+        config.getScaleActionPeriodMillis()
+    );
+  }
+
+  @Override
+  public void stop()
+  {
+    scalingDecisionExec.shutdownNow();
+    metricsCollectionExec.shutdownNow();
+    log.info("CostBasedAutoScaler stopped for dataSource [%s]", dataSource);
+  }
+
+  @Override
+  public void reset()
+  {
+    currentMetrics.set(null);
+  }
+
+  private Runnable collectMetrics()
```

Comment:
Nit: This class may just return `void` and not a `Runnable`. If we need a `Runnable`, we can just the lambda `this::collectMetrics`.